### PR TITLE
[Internal Audit] MsgSplitRouteSwapExactAmountIn & MsgSplitRouteSwapExactAmountOut

### PR DESCRIPTION
**Pull request by stackman27**
Closes: #XXX
https://app.clickup.com/37420681/v/dc/13nzm9-20040/13nzm9-38180
https://app.clickup.com/37420681/v/dc/13nzm9-20020/13nzm9-38160

Added new json based cli commands: 

```

MacBook-Pro-26:osmosis sishirg27$ osmosisd tx poolmanager split-route-swap-exact-amount-out uosmo 1 '{"Route":[{"SwapAmountOutRoute":[{"poolId":"1","TokenInDenom":"uion"}],"TokenOutAmount":1000}]}' --from val --keyring-backend test -b=block --chain-id=localosmosis --fees 10000uosmo

MacBook-Pro-26:osmosis sishirg27$ osmosisd tx poolmanager split-route-swap-exact-amount-in uosmo 1 '{"Route":[{"SwapAmountInRoute":[{"poolId":"1","TokenOutDenom":"uion"}],"TokenInAmount":1000}]}' --from val --keyring-backend test -b=block --chain-id=localosmosis --fees 10000uosmo


```

### CODE REVIEW

**Review by mattverse — CHANGES_REQUESTED** [review]
Main comment is on using the `SwapAmountInRoute` proto directly instead of re-defining them. Unmarshalling part lgtm. 

Also would be great if you can provide logs of using this cli along!

File: x/poolmanager/client/cli/tx.go [hunk 1]
```diff
@@ -57,29 +87,106 @@ func NewSwapExactAmountOutCmd() (*osmocli.TxCliDesc, *types.MsgSwapExactAmountOu
...
+
+	var splitRouteJSONdata RoutesOut
+	err := json.Unmarshal([]byte(inputJSON), &splitRouteJSONdata)
 	if err != nil {
-		return txf, nil, err
+		return nil, err
 	}
 
-	tokenOutMinAmt, ok := sdk.NewIntFromString(tokenOutMinAmtStr)
+	var splitRouteProto []types.SwapAmountOutSplitRoute
```
[the remark this answers is not in the document]
Comment (stackman27) [reply x/poolmanager/client/cli/tx.go:122]: yup file path is much cleaner :))

File: x/poolmanager/client/cli/tx.go [hunk 2]
```diff
@@ -57,29 +87,106 @@ func NewSwapExactAmountOutCmd() (*osmocli.TxCliDesc, *types.MsgSwapExactAmountOu
...
+		Short:            "split route swap exact amount in",
+		ParseAndBuildMsg: NewMsgNewSplitRouteSwapExactAmountIn,
+	}, &types.MsgSplitRouteSwapExactAmountIn{}
+}
+
+func NewSplitRouteSwapExactAmountOut() (*osmocli.TxCliDesc, *types.MsgSplitRouteSwapExactAmountOut) {
+	return &osmocli.TxCliDesc{
+		Use:              "split-route-swap-exact-amount-out [token-out-denom] [token-in-max-amount] [routes]",
+		Short:            "split route swap exact amount out",
+		ParseAndBuildMsg: NewMsgNewSplitRouteSwapExactAmountOut,
```
[the remark this answers is not in the document]
Comment (stackman27) [reply x/poolmanager/client/cli/tx.go:102]: yessir

File: x/poolmanager/client/cli/tx.go [hunk 3]
```diff
@@ -22,11 +23,40 @@ import (
...
+
+type SwapAmountOutRoute struct {
+	PoolID       string `json:"poolId"`
+	TokenInDenom string `json:"TokenInDenom"`
+}
+
+type SwapAmountInRoute struct {
+	PoolID        string `json:"poolId"`
+	TokenOutDenom string `json:"TokenOutDenom"`
+}
```
[the remark this answers is not in the document]
Comment (stackman27) [reply x/poolmanager/client/cli/tx.go:51]: so i tried doing this and ran into parsing issues. basically the JSON parsing wouldn't read the proto structs

File: x/poolmanager/msg_server_test.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+			s.PrepareBalancerPool()
+			s.PrepareBalancerPool()
+
+			msgServer := poolmanagerKeeper.NewMsgServerImpl(s.App.PoolManagerKeeper)
+
+			// Reset event counts to 0 by creating a new manager.
+			ctx = ctx.WithEventManager(sdk.NewEventManager())
+			s.Equal(0, len(ctx.EventManager().Events()))
+
+			response, err := msgServer.SplitRouteSwapExactAmountIn(sdk.WrapSDKContext(ctx), &types.MsgSplitRouteSwapExactAmountIn{
```
[the remark this answers is not in the document]
Comment (stackman27) [reply x/poolmanager/msg_server_test.go:93]: checking this value seems a lot complex than i thought. You have to check error tolerance and see if it's in bound. It's thoroughly tested here: https://github.com/osmosis-labs/osmosis/blob/main/x/poolmanager/router_test.go#L1851-L1856

I can add test but i dont want to overcomplicate test and also dont just want to copy paste the already tested code. please lmk what you think

File: x/poolmanager/msg_server_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+				Routes:           tc.routes,
+				TokenOutDenom:    tc.tokenOutDenom,
+				TokenInMaxAmount: tc.tokenoutMaxAmount,
+			})
+			if tc.expectedError {
+				s.Require().Error(err)
+				s.Require().Nil(response)
+			} else {
+
+				s.Require().NoError(err)
```
[the remark this answers is not in the document]
Comment (stackman27) [reply x/poolmanager/msg_server_test.go:192]: https://github.com/osmosis-labs/osmosis/pull/5247/files#r1201040212

**Comment by stackman27** [comment]
Thank you @mattverse for your feedbacks. Resolved most of it and is R4RR

File: CHANGELOG.md [hunk 6]
```diff
@@ -69,8 +69,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   * [#5129](https://github.com/osmosis-labs/osmosis/pull/5129) Relax twap record validation in init genesis to allow one of the spot prices to be non-zero when twap error is observed.
   * [#5165](https://github.com/osmosis-labs/osmosis/pull/5165) Improve error message when fully exiting from a pool.
   * [#5187](https://github.com/osmosis-labs/osmosis/pull/5187) Expand `IncentivizedPools` query to include internally incentivized CL pools.
-  * [#5239](https://github.com/osmosis-labs/osmosis/pull/5239) Implement `GetTotalPoolShares` public keeper function for GAMM.
-  
+  * [#5239](https://github.com/osmosis-labs/osmosis/pull/5239) Implement `GetTotalPoolShares` public keeper function for GAMM.
```
Comment (stackman27) [inline CHANGELOG.md:72]:
```suggestion
  * [#5239](https://github.com/osmosis-labs/osmosis/pull/5239) Implement `GetTotalPoolShares` public keeper function for GAMM. 
  
```

File: x/poolmanager/client/cli/tx.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattverse) [reply x/poolmanager/client/cli/tx.go]: I should probably work, they are basically same go structs with json labels, did you use the correct json tags? / Do you maybe have a commit or branch I can help debugging with?

**Comment by mattverse** [comment]
@stackman27 Pushed the fix, also fixed according examples for this, please review the last two commits and see if they make sense!

**Review by mattverse — APPROVED** [review]
LGTM

**Review by p0mvn — APPROVED** [review]
utACK based on earlier review and mostly CLI additions.

Thanks @stackman27

**Pull request merged**